Libraries/LLM/LLMModel.swift

```diff
@@ -1,116 +1,36 @@
-    }
-
-    /// Perform an action on the model and/or tokenizer.  Callers _must_ eval any `MLXArray` before returning as
-    /// `MLXArray` is not `Sendable`.
-    public func perform<R>(_ action: @Sendable (LLMModel, Tokenizer) throws -> R) rethrows -> R {
-        try action(model, tokenizer)
-    }
-}
-
-extension Module {
-
-    /// Compute the number of parameters in a possibly quantized model
-    public func numParameters() -> Int {
-        return leafModules().flattenedValues().map {
-            mod -> Int in
-            if let qlin = mod as? QuantizedLinear {
-                return qlin.scales.size * qlin.groupSize
-            } else if let qemb = mod as? QuantizedEmbedding {
-                return qemb.scales.size * qemb.groupSize
-            } else {
-                return mod.parameters().flattenedValues().reduce(
-                    0,
-                    {
-                        $0 + $1.size
-                    })
-            }
-        }.reduce(0, +)
-    }
+// TODO document
+public protocol LLMModel: LanguageModel, LoRAModel {
 }
 
-/// Interface for all LLM Models
-public protocol LLMModel: Module {
-
-    var vocabularySize: Int { get }
+extension LLMModel {
 
-    func callAsFunction(_ inputs: MLXArray, cache: [KVCache]?) -> MLXArray
+    public func prepare(_ input: LMInput, cache: [KVCache], windowSize: Int?) throws
+        -> PrepareResult
+    {
 
-    /// create a new array of ``KVCache`` -- automatic implementation if self
-    /// implements ``KVCacheDimensionProvider``
-    func newCache(parameters: GenerateParameters) -> [KVCache]
+        let prefillStepSize = windowSize ?? 512
+        var y = input.text
+        var state: LMOutput.State? = nil
 
-    /// Optionally preprocess the weights and modify / remove values as needed.
-    func sanitize(weights: [String: MLXArray]) -> [String: MLXArray]
-}
-
-/// Optional protocol that can be implemented by ``LLMModel`` and will
-/// provide an automatic implementation of ``LLMModel/newCache(parameters:)``
-public protocol KVCacheDimensionProvider {
-    var kvHeads: [Int] { get }
-    var headDim: IntOrPair { get }
-}
+        // prepare the prompt in chunks if larger than the prefill size
+        while y.tokens.size > prefillStepSize {
+            let input = y[.newAxis, ..<prefillStepSize]
+            let result = self(input, cache: cache.isEmpty ? nil : cache, state: state)
+            eval(cache)
+            y = y[prefillStepSize...]
+        }
 
-extension LLMModel {
-    public func sanitize(weights: [String: MLXArray]) -> [String: MLXArray] {
-        weights
+        return .tokens(y)
     }
 }
 
-extension LLMModel where Self: KVCacheDimensionProvider {
-    public func newCache(parameters: GenerateParameters) -> [KVCache] {
-        kvHeads.map { n in
-            KVCacheSimple(headDim: headDim, kvHeads: n)
-        }
-    }
-}
+// TODO move? these cause some ambiguity -- how to resolve?
```

Comment:
I was playing around with these to avoid breaking API -- moving types into LMCommon means callers will need to import LMCommon if they refer to them.  This (the aliases) caused more trouble than I think it is worth